### PR TITLE
Reimplement problem transferring and emitting from the worker-side

### LINKED ISSUE #27213: Fix message ordering when reporting problems from a worker
Tests like [`problems are logged worker lifecycle is logged in #isolationMode`](https://github.com/gradle/gradle/blob/cc2a5bba6801a74c8167b7cda14a6df3e90a6065/platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorProblemsApiIntegrationTest.groovy#L68-L98) were unexpectedly flaky.
When having a deeper look, we realized that [`BuildOperationTrace`](https://github.com/gradle/gradle/blob/41ee1247ffc6c1575aa0cfe5ffd38c38c8410fd6/subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java) causes the issues as we broke an unchecked rule:  _operation detail messages should arrive between the start and the stop_.

Sounds reasonable, but our implementation thinks otherwise. Turns out that each outgoing object you add, e.g. where we currently register outgoing messages for problem reporting: 
https://github.com/gradle/gradle/blob/54359d231525e40416e9c852153dafeffe1cf575/platforms/core-execution/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java#L187-L193
All of these `addOutgoing` calls will create a separate thread for each `Handler`:
https://github.com/gradle/gradle/blob/3bc6197f3c80d839ce09ff15bed9da817bcb2a4f/platforms/core-runtime/messaging/src/main/java/org/gradle/internal/remote/internal/hub/MessageHub.java#L133

 The start/stop is coming from [`ResponseProtocol`](https://github.com/gradle/gradle/blob/ceb0244fa27dbe8f60d2c2dc20d8ab8f21e2307a/subprojects/core/src/main/java/org/gradle/process/internal/worker/request/ResponseProtocol.java), and problem reporting comes from [`WorkerProblemProtocol`](https://github.com/gradle/gradle/blob/54359d231525e40416e9c852153dafeffe1cf575/platforms/core-execution/worker-processes/src/main/java/org/gradle/process/internal/worker/problem/WorkerProblemProtocol.java). This causes two different threads racing with each other, and by this, providing zero guarantee of the problem message arriving _before_ the stop message.

We came to the conclusion that the _queue per type_ semantics makes sense, and we should add the problem reporting to the `ResponseProtocol`. Maybe we should rename the class to signify better that it's handling the communication _per work action_.
```[tasklist]
### Actions
- [ ] https://github.com/gradle/gradle/pull/27215
- [ ] #27222
- [ ] Make a PR fixing the issue, and re-enabling the test
```

### PR DESCRIPTION
**Pull request by hegyibalint**
Fixes #27213

I've started to add some documentation (especially class JavaDocs) where I started to grasp what's happening.

@ghale if you have some time, I would be glad if you would take a look at this. Especially at the added comments if they are right.

### CODE REVIEW

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt
Comment (hegyibalint) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:4902]: These disappeared, as I've turned the respective code using lambdas instead of anonymous classes.

File: subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt
Comment (hegyibalint) [inline subprojects/architecture-test/src/changes/archunit_store/internal-api-nullability.txt:4903]: These disappeared, as I've turned the respective code using lambdas instead of anonymous classes.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java
Comment (hegyibalint) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java:53]: Originally the logging was implemented by inheritance (this class extended the default logging implementation)
This didn't feel right after adding the problem reporting and turning this into a composition-based solution instead of inheritance-based one.

**Review by donat — CHANGES_REQUESTED** [review]
Looks good from our side, but I'd definitely wait for @ghale to give a final green light.

One thing that still bothers me is that we implicitly rely on the fact that DefaultProblem carries the current build operation ID. I think it should either
- Report the id in the protocol (i.e.`WorkerProblemProtocol.reportProblem(Problem, BuildOperationId)`), or
- Store the operation id in a wrapper object and unwrap it on the daemon side (i.e. Implement the `WorkerProblem` class with `getProblem()` and `getOperationId()` methods. and implement `WorkerProblemProtocol.reportProblem(WorkerProblem)`).

We should 💯 strive for a clean `DefaultProblem` data class.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java:53]: 💭 I would even argue that this could be _all_ composition.  i.e. a separate delegate for all of the functions of `ResponseProtocol` so that this class is really just a fanout for those functions.  I don't think you need to change this (unless you want to) - I'm just sharing the thought that it might make it more clear that the aggregation of these things in `ResponseProtocol` is an implementation detail rather than any true relationship between the functions.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/ResponseProtocol.java
Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/request/ResponseProtocol.java:26]: 💅 We might also add some rationale for why these things are a part of `ResponseProtocol`.  (i.e. because these protocols are tied to the build operation whose lifecycle is tightly coupled to the request/response interaction). We do have part of an explanation in `WorkerAction`, but it might make sense to include an explanation here as well.  Just so the next person who stumbles on this little bit of subtlety will have something to help understand it.

File: platforms/core-execution/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java
Comment (ghale) [inline platforms/core-execution/worker-processes/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java:189]: ❔ Is it the case that we will _never_ emit problems outside of an item of work?  I assume so, given that this is removed, but just making sure...

**Review by ghale — CHANGES_REQUESTED** [review]
LGTM. I left a few questions and comments.  I think the only thing we _should_ do before merging is add some more detail in the comments on `ResponseProtocol` but everything else is just comments.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java
Comment (hegyibalint) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java:53]: I originally wanted to do more but scaled back my efforts; I was afraid a larger change would delay this feature, which we really want to be available in 8.6.

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/ResponseProtocol.java
Comment (hegyibalint) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/ResponseProtocol.java:26]: Added, PTAL!

**Review by ghale — APPROVED** [review]

**Comment by hegyibalint** [comment]
@donat is not online, but we agreed that we address his concerns out of this PR (#27366)

**Comment by hegyibalint** [comment]
@bot-gradle merge

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java
Comment (hegyibalint) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java:42]: Added, as the `ResponseProtocol` became also `@NonNullApi`.

Comment (donat) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java:42]: 👍

**Review by donat — APPROVED** [review]

**Comment by hegyibalint** [comment]
@bot-gradle merge

**Comment by hegyibalint** [comment]
@bot-gradle cancel

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

**Pull request merged**
